### PR TITLE
Image transformation chain reimplementation

### PR DESCRIPTION
**Pull request by rexxars**
TL;DR: I found a way to optimize resizing of JPEG-images significantly. Many changes had to be done in order to get this to work.
## Background

I was benchmarking Imagick against a different image backend and turned to Imbo to see how we had implemented the `jpeg:size`-flag, and was surprised to find that it didn't actually give us any speed increase. After a few minutes, I quickly realized that this is caused by us not setting the flag prior to loading the image blob into Imagick.

Obviously I wanted to fix this in Imbo, but it turned out to be quite a lot of work:
- I needed a way to get the size of the smallest input image the transformation chain could accept. This is already done in the image variations listener, but I never liked how it has transformation names hardcoded into it, making it impossible for third-party resize transformations to hook onto this.
- There was no simple way to get the class associated with a given transformation name. You had to simply trigger an event and see if anything picked it up. To make things worse, if multiple transformations of the same name was in the same chain, it was impossible to see which one returned the smallest size.
- Some transformations require the full image, and some transformations alter the image in a way that makes it hard or impossible for subsequent transformations to return the correct minimum input size. For instance, rotating an image 37 degrees and applying a `maxSize`-transformation after this is not resizing just the actual image, but instead a rather weird canvas.
## Changes

To try and summarize the changes done in this branch:
- Imbo now has a `TransformationManager`, which keeps track of all the available transformations. Transformations are no longer configured in the `eventListeners` configuration section, instead they are now in a separate `transformations` section.
- The manager is what triggers the transformations, and should in theory be able to pass on more information to each transformation if we want it to. For instance, some transformations might want to know if the image has been rotated or resized in order to correctly calculate where to draw objects (`drawPois` being one of them)
- Two new interfaces have been added that each transformation can implement:
  - `InputSizeConstraint` - will signal to the manager that this transformation needs a certain input size. Typically this is for transformations that do resizing - `maxSize`, `thumbnail` and `resize` all implement it. It can also be used to indicate that it needs the full size image, that the result of the transformation is unpredictable and should stop the minimum size resolving and can also tell Imbo if any rotation was performed. It also contains a method (`adjustParameters()`) that transformations can implement to adjust any parameters that they use, in case the input size was changed. Currently this is used for `canvas`, `border`, `watermark` and `crop`, which all use coordinates or width/height parameters.
  - `RegionExtractor` - will signal to the manager that the transformation extracts a portion of the image. This is used in combination with the minimum input size resolving in order to find the smallest possible size. `crop` currently implements this.
- Imagick asks the transformation manager what the minimum input size is and sets the `jpeg:size` hint before loading the image blob. This allows libjpeg to use a "shrink-on-load" implementation, which significantly increases the speed of the resizing.
- The image variations listener uses the transformation manager in order to resolve minimum variation size and to adjust transformation parameters. This decouples it from core and enables third-party transformations to benefit from it.
## Result

To benchmark the old implementation against the new, I wrote a script that:
- Queries for 300 images (all 2000+ pixels in width)
- Loops through each image and generates an image url with a `maxSize` transformation set to 523x315
  - Image variations are generated for all the images, closest size to 523px being 1000px
- Downloads the image URLs
- Iterates over the same image set 3 times to remove any outliers

This is all done against a server running locally, so network conditions shouldn't have any say. I've reproduced the results multiple times with just small deviations. The result of resizing 900 images (300 \* 3 iterations):

| Implementation | Variations enabled | Speed |
| --- | --- | --- |
| Old | No | 123.01s |
| New | No | 45.57s |
| Old | Yes | 22.30s |
| New | Yes | 16.73s |
## Conclusions

I think the results speak for themselves. Most importantly, this is a "free" optimization - it doesn't require any more disk space or configuration to set up. Obviously, this only applies to JPEG-images that is able to use the "shrink-on-load" trick, but I'm guessing most peoples image collection consists primarily of JPEGs anyway.
## Further work

First and foremost, the `smartSize` transformation has not been optimized. I think @kbrabrand should do this, being the author of it, and it might be that the transformation manager will require some small changes for it to work with this transformation. 

We should also give this some more testing in a production-like environment. Many combined transformations and order of these might provoke bugs, so we should also try quite a bit of transformation combinations before merging.

### CODE REVIEW

**Comment by kbrabrand** [comment]
This is awesome! :clap: I've added a comment with a link to a gist in #393. Will try to find time to look at getting this optimization to work with `smartSize` soon.

**Comment by christeredvartsen** [comment]
Really nice work @rexxars! The way the transformations are handled here enables more interaction between the transformations which is a nice addition.

File: library/Imbo/EventListener/Imagick.php [hunk 1]
```diff
@@ -82,8 +86,27 @@ public function readImageBlob(EventInterface $event) {
             $image = $event->getResponse()->getModel();
         }
 
+        $shouldOptimize = $jpegSizeHintEnabled && !$event->hasArgument('skipOptimization');
+        if ($shouldOptimize && $eventName === 'image.loaded') {
+            // See if we can hint to imagick that we expect a smaller output
+            $minSize = $event->getTransformationManager()->getMinimumImageInputSize($event);
+            if ($minSize) {
+                $inputSize = $minSize['width'] . 'x' . $minSize['height'];
+                $this->imagick->setOption('jpeg:size', $inputSize);
```
Comment (aliasaria) [inline library/Imbo/EventListener/Imagick.php]: I might be mistaken that this is an issue, but as I read it, the ImageMagick docs state that you should set the jpeg:size to twice the minimum size to avoid artifacts here:
http://www.imagemagick.org/Usage/formats/#jpg_read

"Note that this modifier causes the JPEG library to skip the reading of whole columns and rows of pixels. As such it will produce effects much like the Sampling Resize Operator, including its strong Aliasing Artefacts.

Because of this is it recommended that you specify at least double the final 'resize' of the image, to avoid this problem, just as shown in the above example."

Comment (rexxars) [inline library/Imbo/EventListener/Imagick.php]: Thanks for the input, Ali!

This is definitely something that I've overlooked. I personally couldn't see any difference, but then again my eyesight might not be what it once was ;-)

I guess we could simply set a multiplication factor in the configuration, multiply the minimum size by that, then use the smallest of the calculated size and the original.

Thanks for reporting!

**Comment by matslindh** [comment]
Since `thumbnail` in the original imagemagick code uses a 5x for this operator, so we might want larger than 2x as a default value (my vote would be for 3x, since that would match a 3x3 core around the pixel to be resized, which usually give a good result with bilinear resizing). Just using the hint should be enough, I'm fairly sure imagick will pick the most suitable of the original size and the `jpeg:size` setting anyway.

I'm just not sure if we really need to make this configurable at all right now - another setting to maintain and keep supporting in the future.

File: library/Imbo/EventListener/Imagick.php (hunk 1 above)
Comment (matslindh) [inline library/Imbo/EventListener/Imagick.php]: Since thumbnail in the original imagemagick code uses a 5x for this operator, so we might want larger than 2x as a default value (my vote would be for 3x, since that would match a 3x3 core around the pixel to be resized, which usually give a good result with bilinear resizing). Just using the hint should be enough, I'm fairly sure imagick will pick the most suitable of the original size and the jpeg:size setting anyway.

I'm just not sure if we really need to make this configurable at all right now - another setting to maintain and keep supporting in the future.

Comment (rexxars) [inline library/Imbo/EventListener/Imagick.php]: I'd like to benchmark the difference, at least. For some use cases, the quality loss might be acceptable in favor of a more performant resize operation. If the speed improvement is marginal, I agree with @matslindh's suggestion of dropping the configuration flag.

**Comment by christeredvartsen** [comment]
@rexxars Care to do a rebase against develop?

@kbrabrand Could you see if you might find some time to implement support for the `smartSize` transformation as well?

**Comment by rexxars** [comment]
Rebased.

**Comment by kbrabrand** [comment]
I'll try, but it's been really busy lately with courses at university and the new job :/

File: src/Image/TransformationManager.php [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+            }
+
+            if ($handler instanceof InputSizeConstraint) {
+                $minSize = $handler->setImage($image)->getMinimumInputSize($params, $inputSize);
+
+                // Transformations can return `null` if no transformation will occur,
+                // or `false` if we should stop the minimum input size resolving because
+                // the transformation can't be predicted or calculated in advance
+                // (for instance if it needs access to the Imagick instance)
+                if ($minSize === null) {
```
Comment (christeredvartsen) [inline src/Image/TransformationManager.php]: Should these return values be changed to constants to make it more clear what they mean? Something like:

- `TransformationManager::NO_TRANSFORMATION`
- `TransformationManager::STOP_RESOLVING`

or something similar...

File: src/Image/TransformationManager.php [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+        }
+
+        return [
+            'width'  => (int) ceil($minimum['width']),
+            'height' => (int) ceil($minimum['height']),
+            'index'  => $minimum['index'],
+        ];
+    }
+
+    public function adjustImageTransformations(EventInterface $event) {
```
Comment (christeredvartsen) [inline src/Image/TransformationManager.php]: Missing docblock

File: config/config.default.php [hunk 4]
```diff
@@ -109,6 +109,28 @@
...
+         * loading the image into Imagick, which set a hint to libjpeg that enables
+         * "shrink-on-load", which significantly increases speed or resizing.
+         *
+         * Tradeoffs: Transformations have to adjust parameters based on new input size,
+         * some parameters will be one pixel off. Image quality should be the same for
+         * most images, but there is always the possibility of slightly worse quality
+         *
+         * @var boolean
+         */
+        'jpegSizeHint' => true,
```
Comment (matslindh) [inline config/config.default.php]: I feel that true by default is A-OK. This won't be an issue for most users, and the constraint on image delivery (meaning that the delivered image isn't suddenly one pixel larger or smaller than what the request asks for) is the most important size limitation.

File: src/Application.php [hunk 5]
```diff
@@ -81,6 +82,12 @@ public function run(array $config) {
         // Create a router based on the routes in the configuration and internal routes
         $router = new Router($config['routes']);
 
+        // Create a new image transformation manager
+        $transformationManager = new TransformationManager();
+        if (isset($config['transformations']) && is_array($config['transformations'])) {
```
Comment (matslindh) [inline src/Application.php]: Should we have an explicit error if the parameter isn't an array? We don't have any interface defined that an object could implement, so this would currently be a quiet fail. Replacing `isset` with `!empty()` would allow for ` = false;` and ` = null;` in the config to still work, as long as the default value is `[]` (if it's not - we should have the truth check separate from the empty/array check).

File: src/Image/Transformation/Resize.php [hunk 6]
```diff
@@ -21,34 +20,62 @@
...
+            'height' => $this->image->getHeight(),
+        ]);
+
+        // Fall back if there is no need to resize
+        if (!$size) {
+            return;
+        }
+
+        try {
+            $this->imagick->thumbnailImage($size['width'], $size['height']);
```
Comment (matslindh) [inline src/Image/Transformation/Resize.php]: Overextending try/catch

File: src/Image/Transformation/Rotate.php [hunk 7]
```diff
@@ -69,4 +55,22 @@ public function transform(EventInterface $event) {
...
+     */
+    public function getMinimumInputSize(array $params, array $imageSize) {
+        if (empty($params['angle'])) {
+            throw new TransformationException('Missing required parameter: angle', 400);
+        }
+
+        // If the angle of the rotation is dividable by 90, we can calculate the input
+        // size for the transformation that follow. Otherwise, this will be hard, so we
+        // return false to signal that we can't make any assumptions from this point on
+        if ($params['angle'] % 90 === 0) {
```
Comment (matslindh) [inline src/Image/Transformation/Rotate.php]: Rotating by a value not divisble by 90 is probably not used much in the wild anyway, but we should note this as a future enhancement (it wouldn't be too hard to calculate the bounding box of a rotation, but it's not worth delaying the patch further for it.

File: src/Image/Transformation/Resize.php (hunk 6 above)
Comment (matslindh) [inline src/Image/Transformation/Resize.php]: These happen in multiple locations in the transformations - we should probably use everything not relevant for the exception out of the try/catch, but it'd require changes in multiple locations.

File: src/Image/Transformation/Transformation.php [hunk 8]
```diff
@@ -63,9 +105,51 @@ protected function getQuantumRange() {
...
+
+    /**
+     * Get the imagick version
+     *
+     * @return string
+     */
+    protected function getImagickVersion() {
+        // Newer versions of Imagick expose getVersion as a static method,
+        // and won't allow phpunit to mock it even when called on an instance
+        if (method_exists('Imagick', 'getVersion')) {
```
Comment (matslindh) [inline src/Image/Transformation/Transformation.php]: :+1: But let's create a separate issue for that and allow this if it works for now. We have a least one other code path that has to check the ImageMagick version.

File: tests/phpunit/unit/EventListener/ImageVariationsTest.php [hunk 9]
```diff
@@ -596,11 +553,15 @@ public function testGenerateVariationsIncludesSpecifiedWidths() {
     public function testGenerateVariationsTriggersWarningOnTransformationException() {
         $this->imageModel->method('getWidth')->willReturn(1024);
 
-        $this->eventManager->expects($this->at(1))
-            ->method('trigger')
-            ->with('image.transformation.resize', $this->anything())
+        $this->transformation = $this->createMock('Imbo\Image\Transformation\Transformation');
+        $this->transformation->expects($this->any())->method('setImage')->will($this->returnSelf());
+        $this->transformation->expects($this->at(1))
```
Comment (matslindh) [inline tests/phpunit/unit/EventListener/ImageVariationsTest.php]: This is tied to the implementation of the transformation chain (.. even more than what a mock already is)

File: tests/phpunit/unit/EventListener/MaxImageSizeTest.php [hunk 10]
```diff
@@ -74,9 +74,18 @@ public function testWillTriggerTransformationWhenImageIsAboveTheLimits($imageWid
...
+            $maxSize->expects($this->once())->method('transform')->with(['width' => $maxWidth, 'height' => $maxHeight]);
+
+            $transformationManager = $this->createMock('Imbo\Image\TransformationManager');
+            $transformationManager
+                ->expects($this->once())
+                ->method('getTransformation')
+                ->with('maxSize')
+                ->will($this->returnValue($maxSize));
+
+            $event->expects($this->once())->method('getTransformationManager')->will($this->returnValue($transformationManager));
```
Comment (matslindh) [inline tests/phpunit/unit/EventListener/MaxImageSizeTest.php]: Multiple uses of `->once()` tied to implementation in this block

File: tests/phpunit/unit/Http/Response/ResponseFormatterTest.php [hunk 11]
```diff
@@ -348,17 +348,20 @@ public function testTriggersAConversionTransformationIfNeededWhenTheModelIsAnIma
...
+        $convert->expects($this->once())->method('transform')->with(['type' => 'png']);
+
+        $transformationManager = $this->createMock('Imbo\Image\TransformationManager');
+        $transformationManager
+            ->expects($this->once())
+            ->method('getTransformation')
+            ->with('convert')
+            ->will($this->returnValue($convert));
+
+        $this->event->expects($this->once())->method('getTransformationManager')->will($this->returnValue($transformationManager));
```
Comment (matslindh) [inline tests/phpunit/unit/Http/Response/ResponseFormatterTest.php]: Multiple uses of `->once()` in block

File: tests/phpunit/unit/Image/Transformation/AutoRotateTest.php [hunk 12]
```diff
@@ -28,11 +29,25 @@ public function testWillNotUpdateTheImageWhenNotNeeded() {
...
+    /**
+     * @covers Imbo\Image\Transformation\AutoRotate::transform
+     */
+    public function testWillRotateWhenNeeded() {
+        $imagick = $this->createMock('Imagick');
+        $imagick->expects($this->once())->method('getImageOrientation')->will($this->returnValue(
+            Imagick::ORIENTATION_TOPRIGHT
+        ));
+        $imagick->expects($this->once())->method('flopImage');
+        $imagick->expects($this->once())->method('setImageOrientation')->with(Imagick::ORIENTATION_TOPLEFT);
```
Comment (matslindh) [inline tests/phpunit/unit/Image/Transformation/AutoRotateTest.php]: At least a few of these should be changed to `->any()` (the flopImage check _might_ be allowed a `once()`, the other two should be `any()`.

File: tests/phpunit/unit/Image/Transformation/StripTest.php [hunk 13]
```diff
@@ -31,6 +31,19 @@ public function testThrowsCorrectExceptionWhenAnErrorOccurs() {
...
+        $image = $this->createMock('Imbo\Model\Image');
+        $image->expects($this->once())->method('hasBeenTransformed')->with(true);
+
+        $imagick = $this->createMock('Imagick');
+        $imagick->expects($this->once())->method('getImageBlob')->will($this->returnValue('foo'));
+        $imagick->expects($this->once())->method('clear');
+        $imagick->expects($this->once())->method('readImageBlob')->with('foo');
+
+        $transformation = new Strip();
+        $transformation->setImagick($imagick)->setImage($image)->transform([]);
```
Comment (matslindh) [inline tests/phpunit/unit/Image/Transformation/StripTest.php]: Multiple uses of `once()`

File: tests/phpunit/unit/Image/Transformation/SmartSizeTest.php [hunk 14]
```diff
@@ -148,24 +148,28 @@ public function testSmartSize($imageDimensions, $params, $cropParams) {
...
+        $headerBag->expects($this->once())->method('set')->with('X-Imbo-POIs-Used', 1);
+
         $response = new Response();
+        $response->headers = $headerBag;
 
         $event = $this->createMock('Imbo\EventManager\Event');
-        $event->expects($this->at(0))->method('getArgument')->with('image')->will($this->returnValue($image));
-        $event->expects($this->at(1))->method('getArgument')->with('params')->will($this->returnValue($params));
-        $event->expects($this->at(2))->method('getResponse')->will($this->returnValue($response));
+        $event->expects($this->once())->method('getResponse')->will($this->returnValue($response));
```
Comment (matslindh) [inline tests/phpunit/unit/Image/Transformation/SmartSizeTest.php]: Multiple uses of once

File: tests/phpunit/unit/Image/Transformation/WatermarkTest.php [hunk 15]
```diff
@@ -73,14 +67,13 @@ public function testThrowsExceptionIfSpecifiedImageIsNotFound() {
...
 
         $event = $this->createMock('Imbo\EventManager\Event');
-        $event->expects($this->at(0))->method('getArgument')->with('image')->will($this->returnValue($image));
-        $event->expects($this->at(1))->method('getArgument')->with('params')->will($this->returnValue([
-            'img' => 'foobar',
-        ]));
-        $event->expects($this->at(2))->method('getStorage')->will($this->returnValue($storage));
-        $event->expects($this->at(3))->method('getRequest')->will($this->returnValue($request));
+        $event->expects($this->once())->method('getStorage')->will($this->returnValue($storage));
+        $event->expects($this->once())->method('getRequest')->will($this->returnValue($request));
```
Comment (matslindh) [inline tests/phpunit/unit/Image/Transformation/WatermarkTest.php]: Multiple uses of once()

File: src/Application.php (hunk 5 above)
Comment (christeredvartsen) [inline src/Application.php]: > Should we have an explicit error if the parameter isn't an array?

I think so. Will fix.
